arch/riscv/core/thread.c

```diff
@@ -133,3 +179,96 @@ int arch_float_enable(struct k_thread *thread)
 	return 0;
 }
 #endif /* CONFIG_FPU && CONFIG_FPU_SHARING */
+
+#ifdef CONFIG_USERSPACE
+
+/* Function used by Zephyr to switch a supervisor thread to a user thread */
+FUNC_NORETURN void arch_user_mode_enter(k_thread_entry_t user_entry,
+	void *p1, void *p2, void *p3)
+{
+	arch_syscall_invoke5((uintptr_t) arch_user_mode_enter,
+		(uintptr_t) user_entry,
+		(uintptr_t) p1,
+		(uintptr_t) p2,
+		(uintptr_t) p3,
+		FORCE_SYSCALL_ID);
+
+	CODE_UNREACHABLE;
+}
+
+/*
+ * User space entry function
+ *
+ * This function is the entry point to user mode from privileged execution.
+ * The conversion is one way, and threads which transition to user mode do
+ * not transition back later, unless they are doing system calls.
+ */
+FUNC_NORETURN void z_riscv_user_mode_enter_syscall(k_thread_entry_t user_entry,
+	void *p1, void *p2, void *p3)
+{
+	ulong_t top_of_user_stack = 0U;
+	uintptr_t status;
+
+	/* Set up privileged stack */
+#ifdef CONFIG_GEN_PRIV_STACKS
+		_current->arch.priv_stack_start =
+			(uint32_t)z_priv_stack_find(_current->stack_obj);
+#else
+		_current->arch.priv_stack_start =
+			(uint32_t)(_current->stack_obj);
+#endif /* CONFIG_GEN_PRIV_STACKS */
+
+	top_of_user_stack = Z_STACK_PTR_ALIGN(
+				_current->stack_info.start +
+				_current->stack_info.size -
+				_current->stack_info.delta);
+
+	/* Set next CPU status to user mode */
+	status = csr_read(mstatus);
+	status = INSERT_FIELD(status, MSTATUS_MPP, PRV_U);
+	status = INSERT_FIELD(status, MSTATUS_MPRV, 0);
+
+	csr_write(mstatus, status);
+	csr_write(mepc, z_thread_entry_wrapper);
+
+	/* Set up Physical Memory Protection */
+#if defined(CONFIG_PMP_STACK_GUARD)
+	z_riscv_init_stack_guard(_current);
+#endif
+
+	z_riscv_init_user_accesses(_current);
+	z_riscv_configure_user_allowed_stack(_current);
+
+	is_user_mode = true;
+
+	__asm__ volatile ("mv sp, %1"
```

Comment:
@amergnat There is a subtle bug here. If your lucky (as is the case with the current SDK compiler) this will work fine, however, its possible that the compiler saves the function parameters (user_entry, p0, p1, p2, p3) to the stack and then attempts to restore them below (as is the case with SiFive's released toolchain). This will load invalid values since the sp has been changed here. I think the best way to avoid this is to set the sp last before the mret instruction.

Reply:
Ok, it will be fixed